apps/emqx_authn/src/emqx_authn_utils.erl

```diff
@@ -100,41 +100,52 @@ check_password_from_selected_map(Algorithm, Selected, Password) ->
     end.
 
 parse_deep(Template) ->
-    emqx_placeholder:preproc_tmpl_deep(Template, #{placeholders => ?AUTHN_PLACEHOLDERS}).
+    Result = emqx_connector_template:parse_deep(Template),
+    ok = emqx_connector_template:validate(?AUTHN_PLACEHOLDERS, Result),
```

Comment:
Q: Also, here's the potential backwards compatibility breakage: any authn configurations with typos in templates won't be created. Again, my reason for that is that loudly failing when dealing with security-related configuration is better. Yet, I'm not sure if we're ready to introduce such an incompatibility.

Reply:
It seems a bit unfriendly to fail with `badmatch` on validation errors. Mb introduce special errors? `error({validation_error, {disallowed_vars, [...]})` or smth.

I also do not think we may introduce incompatibilities that may break existing configurations in a minor release or as a part of refactoring. @zmstone ?

As far as I remember some previous discussions, we tend to choose compatibility over enforcing security (like stricter JWT validations, etc.). I would suggest introducing an additional (safer) API and switching to it separately.

Reply:
Thanks for the input. Followed the compatibility route in de879510f553ec10a88dfb24cd26c5187e2294e0 (there are now warnings though when users give us templates with what look like typos).